plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcModule.java

```diff
@@ -66,11 +68,17 @@ public void configure(Binder binder)
         bindSessionPropertiesProvider(binder, JdbcMetadataSessionProperties.class);
 
         binder.bind(JdbcClient.class).to(CachingJdbcClient.class).in(Scopes.SINGLETON);
-        binder.bind(ConnectionFactory.class).to(Key.get(ConnectionFactory.class, StatsCollecting.class));
 
         newOptionalBinder(binder, Key.get(int.class, MaxDomainCompactionThreshold.class));
     }
 
+    @Provides
+    @Singleton
+    public ConnectionFactory createConnectionFactory(@StatsCollecting ConnectionFactory connectionFactory)
+    {
+        return new LazyConnectionFactory(connectionFactory);
```

Comment:
There is no point in wrapping it for everyone.
Whether laziness makes any sense depends on the usage pattern and so it's limited to JDBC client.

Why not wrap it there, eg on https://github.com/trinodb/trino/blob/5f76a07ec6545e1be96de821a15263fc73c11f85/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java#L159 ?